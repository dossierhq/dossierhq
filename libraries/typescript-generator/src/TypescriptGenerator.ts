import type {
  AdminEntityTypeSpecification,
  AdminFieldSpecification,
  AdminSchema,
  AdminValueTypeSpecification,
  PublishedEntityTypeSpecification,
  PublishedFieldSpecification,
  PublishedSchema,
  PublishedValueTypeSpecification,
} from '@jonasb/datadata-core';
import { FieldType } from '@jonasb/datadata-core';

interface GeneratorContext {
  coreImports: Set<string>;
}

export function generateTypescriptForSchema({
  adminSchema,
  publishedSchema,
}: {
  adminSchema: AdminSchema | null;
  publishedSchema: PublishedSchema | null;
}) {
  const context: GeneratorContext = { coreImports: new Set<string>() };
  const paragraphs: string[] = [];

  if (adminSchema) {
    for (const entitySpec of adminSchema.spec.entityTypes) {
      paragraphs.push(...generateAdminEntityType(context, entitySpec));
    }
    paragraphs.push(...generateAllTypesUnion(adminSchema.spec.valueTypes, 'Admin', 'ValueItems'));
    for (const valueSpec of adminSchema.spec.valueTypes) {
      paragraphs.push(...generateAdminValueType(context, valueSpec));
    }
  }
  if (publishedSchema) {
    for (const entitySpec of publishedSchema.spec.entityTypes) {
      paragraphs.push(...generatePublishedEntityType(context, entitySpec));
    }
    paragraphs.push(
      ...generateAllTypesUnion(publishedSchema.spec.valueTypes, 'Published', 'ValueItems')
    );
    for (const valueSpec of publishedSchema.spec.valueTypes) {
      paragraphs.push(...generatePublishedValueType(context, valueSpec));
    }
  }

  if (context.coreImports.size > 0) {
    const importStatement = `import type { ${[...context.coreImports]
      .sort()
      .join(', ')} } from '@jonasb/datadata-core';`;
    // insert
    paragraphs.splice(0, 0, importStatement);
  }
  paragraphs.push(''); // final newline
  return paragraphs.join('\n');
}

function generateAllTypesUnion(
  types: AdminEntityTypeSpecification[] | PublishedEntityTypeSpecification[],
  adminOrPublished: 'Admin' | 'Published',
  entitiesOrValueItems: 'Entities' | 'ValueItems'
) {
  const typeDefinition =
    types.length === 0 ? 'never' : types.map((it) => `${adminOrPublished}${it.name}`).join(' | ');
  return ['', `export type All${adminOrPublished}${entitiesOrValueItems} = ${typeDefinition};`];
}

function generateAdminEntityType(
  context: GeneratorContext,
  entitySpec: AdminEntityTypeSpecification
) {
  return generateEntityType(context, entitySpec, 'Admin');
}

function generatePublishedEntityType(
  context: GeneratorContext,
  entitySpec: PublishedEntityTypeSpecification
) {
  return generateEntityType(context, entitySpec, 'Published');
}

function generateEntityType(
  context: GeneratorContext,
  entitySpec: PublishedEntityTypeSpecification,
  adminOrPublished: 'Admin' | 'Published'
) {
  const paragraphs: string[] = [''];

  // fields type
  const fieldsName = `${adminOrPublished}${entitySpec.name}Fields`;
  if (entitySpec.fields.length === 0) {
    paragraphs.push(`export type ${fieldsName} = Record<never, never>;`);
  } else {
    paragraphs.push(`export interface ${fieldsName} {`);
    for (const fieldSpec of entitySpec.fields) {
      paragraphs.push(`  ${fieldSpec.name}: ${fieldType(context, fieldSpec, adminOrPublished)};`);
    }
    paragraphs.push(`}`);
  }
  paragraphs.push('');

  // entity type
  const parentTypeName = `${adminOrPublished}Entity`;
  const entityTypeName = `${adminOrPublished}${entitySpec.name}`;
  context.coreImports.add(parentTypeName);
  paragraphs.push(
    `export type ${entityTypeName} = ${parentTypeName}<'${entitySpec.name}', ${fieldsName}>;`
  );

  // isAdminFoo() / isPublishedFoo()
  paragraphs.push('');
  paragraphs.push(
    `export function is${entityTypeName}(entity: ${parentTypeName} | ${entityTypeName}): entity is ${entityTypeName} {`
  );
  paragraphs.push(`  return entity.info.type === '${entitySpec.name}';`);
  paragraphs.push(`}`);

  // assertIsAdminFoo() / assertIsPublishedFoo()
  paragraphs.push('');
  paragraphs.push(
    `export function assertIs${entityTypeName}(entity: ${parentTypeName} | ${entityTypeName}): asserts entity is ${entityTypeName} {`
  );
  paragraphs.push(`  if (entity.info.type !== '${entitySpec.name}') {`);
  paragraphs.push(
    `    throw new Error('Expected info.type = ${entitySpec.name} (but was ' + entity.info.type + ')');`
  );
  paragraphs.push(`  }`);
  paragraphs.push(`}`);

  return paragraphs;
}

function generateAdminValueType(context: GeneratorContext, valueSpec: AdminValueTypeSpecification) {
  return generateValueType(context, valueSpec, 'Admin');
}

function generatePublishedValueType(
  context: GeneratorContext,
  valueSpec: PublishedValueTypeSpecification
) {
  return generateValueType(context, valueSpec, 'Published');
}

function generateValueType(
  context: GeneratorContext,
  valueSpec: PublishedValueTypeSpecification,
  adminOrPublished: 'Admin' | 'Published'
) {
  const paragraphs: string[] = [''];

  // fields type
  const fieldsName = `${adminOrPublished}${valueSpec.name}Fields`;
  if (valueSpec.fields.length === 0) {
    paragraphs.push(`export type ${fieldsName} = Record<never, never>;`);
  } else {
    paragraphs.push(`export interface ${fieldsName} {`);
    for (const fieldSpec of valueSpec.fields) {
      paragraphs.push(`  ${fieldSpec.name}: ${fieldType(context, fieldSpec, adminOrPublished)};`);
    }
    paragraphs.push(`}`);
  }
  paragraphs.push('');

  // value type
  const parentTypeName = 'ValueItem';
  const parentTypeInName = 'ValueItem<string, object>';
  const valueTypeName = `${adminOrPublished}${valueSpec.name}`;
  context.coreImports.add(parentTypeName);
  paragraphs.push(
    `export type ${valueTypeName} = ${parentTypeName}<'${valueSpec.name}', ${fieldsName}>;`
  );

  // isAdminFoo() / isPublishedFoo()
  paragraphs.push('');
  paragraphs.push(
    `export function is${valueTypeName}(valueItem: ${parentTypeInName} | ${valueTypeName}): valueItem is ${valueTypeName} {`
  );
  paragraphs.push(`  return valueItem.type === '${valueSpec.name}';`);
  paragraphs.push(`}`);

  // assertIsAdminFoo() / assertIsPublishedFoo()
  paragraphs.push('');
  paragraphs.push(
    `export function assertIs${valueTypeName}(valueItem: ${parentTypeInName} | ${valueTypeName}): asserts valueItem is ${valueTypeName} {`
  );
  paragraphs.push(`  if (valueItem.type !== '${valueSpec.name}') {`);
  paragraphs.push(
    `    throw new Error('Expected type = ${valueSpec.name} (but was ' + valueItem.type + ')');`
  );
  paragraphs.push(`  }`);
  paragraphs.push(`}`);

  return paragraphs;
}

function fieldType(
  { coreImports }: GeneratorContext,
  fieldSpec: AdminFieldSpecification | PublishedFieldSpecification,
  adminOrPublished: 'Admin' | 'Published'
) {
  let type: string;
  switch (fieldSpec.type) {
    case FieldType.Boolean:
      type = 'boolean';
      break;
    case FieldType.EntityType:
      coreImports.add('EntityReference');
      type = 'EntityReference';
      break;
    case FieldType.Location:
      coreImports.add('Location');
      type = 'Location';
      break;
    case FieldType.RichText:
      coreImports.add('RichText');
      type = 'RichText';
      break;
    case FieldType.String:
      type = 'string';
      break;
    case FieldType.ValueType:
      if (fieldSpec.valueTypes && fieldSpec.valueTypes.length > 0) {
        type = fieldSpec.valueTypes.map((it) => `${adminOrPublished}${it}`).join(' | ');
      } else {
        type = `All${adminOrPublished}ValueItems`;
      }
      break;
  }
  return fieldSpec.list ? `Array<${type}> | null` : `${type} | null`;
}
