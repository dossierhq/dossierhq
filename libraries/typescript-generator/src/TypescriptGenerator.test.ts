import { AdminSchema, FieldType } from '@dossierhq/core';
import { describe, expect, test } from 'vitest';
import { generateTypescriptForSchema } from './TypescriptGenerator.js';

describe('generateTypescriptForSchema', () => {
  test('empty', () => {
    expect(
      generateTypescriptForSchema({
        adminSchema: AdminSchema.createAndValidate({}).valueOrThrow(),
        publishedSchema: null,
      })
    ).toMatchSnapshot();
  });

  test('Foo (no fields)', () => {
    expect(
      generateTypescriptForSchema({
        adminSchema: AdminSchema.createAndValidate({
          entityTypes: [{ name: 'Foo', fields: [] }],
        }).valueOrThrow(),
        publishedSchema: null,
      })
    ).toMatchSnapshot();
  });

  test('Booleans', () => {
    expect(
      generateTypescriptForSchema({
        adminSchema: AdminSchema.createAndValidate({
          entityTypes: [
            {
              name: 'Booleans',
              fields: [
                { name: 'boolean', type: FieldType.Boolean },
                { name: 'booleanList', type: FieldType.Boolean, list: true },
              ],
            },
          ],
        }).valueOrThrow(),
        publishedSchema: null,
      })
    ).toMatchSnapshot();
  });

  test('EntityTypes', () => {
    expect(
      generateTypescriptForSchema({
        adminSchema: AdminSchema.createAndValidate({
          entityTypes: [
            {
              name: 'EntityTypes',
              fields: [
                { name: 'entityType', type: FieldType.Entity },
                { name: 'entityTypeList', type: FieldType.Entity, list: true },
              ],
            },
          ],
        }).valueOrThrow(),
        publishedSchema: null,
      })
    ).toMatchSnapshot();
  });

  test('Locations', () => {
    expect(
      generateTypescriptForSchema({
        adminSchema: AdminSchema.createAndValidate({
          entityTypes: [
            {
              name: 'Locations',
              fields: [
                { name: 'location', type: FieldType.Location },
                { name: 'locationList', type: FieldType.Location, list: true },
              ],
            },
          ],
        }).valueOrThrow(),
        publishedSchema: null,
      })
    ).toMatchSnapshot();
  });

  test('Numbers', () => {
    expect(
      generateTypescriptForSchema({
        adminSchema: AdminSchema.createAndValidate({
          entityTypes: [
            {
              name: 'Numbers',
              fields: [
                { name: 'number', type: FieldType.Number },
                { name: 'numberInteger', type: FieldType.Number, integer: true },
                { name: 'numberList', type: FieldType.Number, list: true },
                { name: 'numberIntegerList', type: FieldType.Number, list: true, integer: true },
              ],
            },
          ],
        }).valueOrThrow(),
        publishedSchema: null,
      })
    ).toMatchSnapshot();
  });

  test('RichTexts', () => {
    expect(
      generateTypescriptForSchema({
        adminSchema: AdminSchema.createAndValidate({
          entityTypes: [
            {
              name: 'RichTexts',
              fields: [
                { name: 'richText', type: FieldType.RichText },
                { name: 'richTextList', type: FieldType.RichText, list: true },
              ],
            },
          ],
        }).valueOrThrow(),
        publishedSchema: null,
      })
    ).toMatchSnapshot();
  });

  test('Strings', () => {
    expect(
      generateTypescriptForSchema({
        adminSchema: AdminSchema.createAndValidate({
          entityTypes: [
            {
              name: 'Strings',
              fields: [
                { name: 'string', type: FieldType.String, required: true },
                { name: 'stringList', type: FieldType.String, list: true, required: true },
              ],
            },
          ],
        }).valueOrThrow(),
        publishedSchema: null,
      })
    ).toMatchSnapshot();
  });

  test('ValueTypes', () => {
    expect(
      generateTypescriptForSchema({
        adminSchema: AdminSchema.createAndValidate({
          entityTypes: [
            {
              name: 'ValueTypes',
              fields: [
                { name: 'valueType', type: FieldType.ValueItem, required: true },
                {
                  name: 'valueTypeFoo',
                  type: FieldType.ValueItem,
                  valueTypes: ['Foo'],
                  required: true,
                },
                {
                  name: 'valueTypeFooBar',
                  type: FieldType.ValueItem,
                  valueTypes: ['Foo', 'Bar'],
                  required: true,
                },
                { name: 'valueTypeList', type: FieldType.ValueItem, list: true, required: true },
                {
                  name: 'valueTypeListFoo',
                  type: FieldType.ValueItem,
                  valueTypes: ['Foo'],
                  list: true,
                  required: true,
                },
                {
                  name: 'valueTypeListFooBar',
                  type: FieldType.ValueItem,
                  valueTypes: ['Foo', 'Bar'],
                  list: true,
                  required: true,
                },
              ],
            },
          ],
          valueTypes: [
            {
              name: 'Foo',
              fields: [{ name: 'string', type: FieldType.String, required: true }],
            },
            {
              name: 'Bar',
              fields: [{ name: 'string', type: FieldType.String, required: true }],
            },
          ],
        }).valueOrThrow(),
        publishedSchema: null,
      })
    ).toMatchSnapshot();
  });

  test('ValueItem (no fields)', () => {
    expect(
      generateTypescriptForSchema({
        adminSchema: AdminSchema.createAndValidate({
          entityTypes: [
            {
              name: 'ValueItem',
              fields: [{ name: 'valueItem', type: FieldType.ValueItem, required: true }],
            },
          ],
          valueTypes: [{ name: 'Foo', fields: [] }],
        }).valueOrThrow(),
        publishedSchema: null,
      })
    ).toMatchSnapshot();
  });

  test('Unique index', () => {
    expect(
      generateTypescriptForSchema({
        adminSchema: AdminSchema.createAndValidate({
          entityTypes: [
            {
              name: 'Strings',
              fields: [{ name: 'string', type: FieldType.String, index: 'slug' }],
            },
          ],
          indexes: [{ name: 'slug', type: 'unique' }],
        }).valueOrThrow(),
        publishedSchema: null,
      })
    ).toMatchSnapshot();
  });
});

describe('generateTypescriptForSchema published', () => {
  test('empty', () => {
    expect(
      generateTypescriptForSchema({
        adminSchema: null,
        publishedSchema: AdminSchema.createAndValidate({}).valueOrThrow().toPublishedSchema(),
      })
    ).toMatchSnapshot();
  });

  test('adminOnly entity type', () => {
    expect(
      generateTypescriptForSchema({
        adminSchema: null,
        publishedSchema: AdminSchema.createAndValidate({
          entityTypes: [{ name: 'Foo', adminOnly: true, fields: [] }],
        })
          .valueOrThrow()
          .toPublishedSchema(),
      })
    ).toMatchSnapshot();
  });

  test('adminOnly field', () => {
    expect(
      generateTypescriptForSchema({
        adminSchema: null,
        publishedSchema: AdminSchema.createAndValidate({
          entityTypes: [
            {
              name: 'Foo',
              fields: [{ name: 'field', adminOnly: true, type: FieldType.String }],
            },
          ],
        })
          .valueOrThrow()
          .toPublishedSchema(),
      })
    ).toMatchSnapshot();
  });

  test('required fields', () => {
    expect(
      generateTypescriptForSchema({
        adminSchema: null,
        publishedSchema: AdminSchema.createAndValidate({
          entityTypes: [
            {
              name: 'Foo',
              fields: [
                { name: 'string', type: FieldType.String, required: true },
                { name: 'stringList', type: FieldType.String, list: true, required: true },
              ],
            },
          ],
        })
          .valueOrThrow()
          .toPublishedSchema(),
      })
    ).toMatchSnapshot();
  });

  test('Unique index', () => {
    expect(
      generateTypescriptForSchema({
        adminSchema: null,
        publishedSchema: AdminSchema.createAndValidate({
          entityTypes: [
            {
              name: 'Strings',
              fields: [{ name: 'string', type: FieldType.String, index: 'slug' }],
            },
          ],
          indexes: [{ name: 'slug', type: 'unique' }],
        })
          .valueOrThrow()
          .toPublishedSchema(),
      })
    ).toMatchSnapshot();
  });
});
