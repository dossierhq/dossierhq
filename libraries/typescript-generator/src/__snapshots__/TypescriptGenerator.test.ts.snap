// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateTypescriptForSchema > Booleans 1`] = `
"import type { AdminClient, AdminEntity, AdminExceptionClient } from '@dossierhq/core';

export type AppAdminClient = AdminClient<AppAdminEntity, AppAdminUniqueIndexes, AppAdminExceptionClient>;

export type AppAdminExceptionClient = AdminExceptionClient<AppAdminEntity, AppAdminUniqueIndexes>;

export type AppAdminUniqueIndexes = never;

export type AppAdminEntity = AdminBooleans;

export interface AdminBooleansFields {
  boolean: boolean | null;
  booleanList: boolean[] | null;
}

export type AdminBooleans = AdminEntity<'Booleans', AdminBooleansFields, string>;

export function isAdminBooleans(entity: AdminEntity<string, object>): entity is AdminBooleans {
  return entity.info.type === 'Booleans';
}

export function assertIsAdminBooleans(entity: AdminEntity<string, object>): asserts entity is AdminBooleans {
  if (entity.info.type !== 'Booleans') {
    throw new Error('Expected info.type = Booleans (but was ' + entity.info.type + ')');
  }
}

export type AppAdminValueItem = never;
"
`;

exports[`generateTypescriptForSchema > EntityTypes 1`] = `
"import type { AdminClient, AdminEntity, AdminExceptionClient, EntityReference } from '@dossierhq/core';

export type AppAdminClient = AdminClient<AppAdminEntity, AppAdminUniqueIndexes, AppAdminExceptionClient>;

export type AppAdminExceptionClient = AdminExceptionClient<AppAdminEntity, AppAdminUniqueIndexes>;

export type AppAdminUniqueIndexes = never;

export type AppAdminEntity = AdminEntityTypes;

export interface AdminEntityTypesFields {
  entityType: EntityReference | null;
  entityTypeList: EntityReference[] | null;
}

export type AdminEntityTypes = AdminEntity<'EntityTypes', AdminEntityTypesFields, string>;

export function isAdminEntityTypes(entity: AdminEntity<string, object>): entity is AdminEntityTypes {
  return entity.info.type === 'EntityTypes';
}

export function assertIsAdminEntityTypes(entity: AdminEntity<string, object>): asserts entity is AdminEntityTypes {
  if (entity.info.type !== 'EntityTypes') {
    throw new Error('Expected info.type = EntityTypes (but was ' + entity.info.type + ')');
  }
}

export type AppAdminValueItem = never;
"
`;

exports[`generateTypescriptForSchema > Foo (no fields) 1`] = `
"import type { AdminClient, AdminEntity, AdminExceptionClient } from '@dossierhq/core';

export type AppAdminClient = AdminClient<AppAdminEntity, AppAdminUniqueIndexes, AppAdminExceptionClient>;

export type AppAdminExceptionClient = AdminExceptionClient<AppAdminEntity, AppAdminUniqueIndexes>;

export type AppAdminUniqueIndexes = never;

export type AppAdminEntity = AdminFoo;

export type AdminFooFields = Record<never, never>;

export type AdminFoo = AdminEntity<'Foo', AdminFooFields, string>;

export function isAdminFoo(entity: AdminEntity<string, object>): entity is AdminFoo {
  return entity.info.type === 'Foo';
}

export function assertIsAdminFoo(entity: AdminEntity<string, object>): asserts entity is AdminFoo {
  if (entity.info.type !== 'Foo') {
    throw new Error('Expected info.type = Foo (but was ' + entity.info.type + ')');
  }
}

export type AppAdminValueItem = never;
"
`;

exports[`generateTypescriptForSchema > Locations 1`] = `
"import type { AdminClient, AdminEntity, AdminExceptionClient, Location } from '@dossierhq/core';

export type AppAdminClient = AdminClient<AppAdminEntity, AppAdminUniqueIndexes, AppAdminExceptionClient>;

export type AppAdminExceptionClient = AdminExceptionClient<AppAdminEntity, AppAdminUniqueIndexes>;

export type AppAdminUniqueIndexes = never;

export type AppAdminEntity = AdminLocations;

export interface AdminLocationsFields {
  location: Location | null;
  locationList: Location[] | null;
}

export type AdminLocations = AdminEntity<'Locations', AdminLocationsFields, string>;

export function isAdminLocations(entity: AdminEntity<string, object>): entity is AdminLocations {
  return entity.info.type === 'Locations';
}

export function assertIsAdminLocations(entity: AdminEntity<string, object>): asserts entity is AdminLocations {
  if (entity.info.type !== 'Locations') {
    throw new Error('Expected info.type = Locations (but was ' + entity.info.type + ')');
  }
}

export type AppAdminValueItem = never;
"
`;

exports[`generateTypescriptForSchema > Numbers 1`] = `
"import type { AdminClient, AdminEntity, AdminExceptionClient } from '@dossierhq/core';

export type AppAdminClient = AdminClient<AppAdminEntity, AppAdminUniqueIndexes, AppAdminExceptionClient>;

export type AppAdminExceptionClient = AdminExceptionClient<AppAdminEntity, AppAdminUniqueIndexes>;

export type AppAdminUniqueIndexes = never;

export type AppAdminEntity = AdminNumbers;

export interface AdminNumbersFields {
  number: number | null;
  numberInteger: number | null;
  numberList: number[] | null;
  numberIntegerList: number[] | null;
}

export type AdminNumbers = AdminEntity<'Numbers', AdminNumbersFields, string>;

export function isAdminNumbers(entity: AdminEntity<string, object>): entity is AdminNumbers {
  return entity.info.type === 'Numbers';
}

export function assertIsAdminNumbers(entity: AdminEntity<string, object>): asserts entity is AdminNumbers {
  if (entity.info.type !== 'Numbers') {
    throw new Error('Expected info.type = Numbers (but was ' + entity.info.type + ')');
  }
}

export type AppAdminValueItem = never;
"
`;

exports[`generateTypescriptForSchema > RichTexts 1`] = `
"import type { AdminClient, AdminEntity, AdminExceptionClient, RichText } from '@dossierhq/core';

export type AppAdminClient = AdminClient<AppAdminEntity, AppAdminUniqueIndexes, AppAdminExceptionClient>;

export type AppAdminExceptionClient = AdminExceptionClient<AppAdminEntity, AppAdminUniqueIndexes>;

export type AppAdminUniqueIndexes = never;

export type AppAdminEntity = AdminRichTexts;

export interface AdminRichTextsFields {
  richText: RichText | null;
  richTextList: RichText[] | null;
}

export type AdminRichTexts = AdminEntity<'RichTexts', AdminRichTextsFields, string>;

export function isAdminRichTexts(entity: AdminEntity<string, object>): entity is AdminRichTexts {
  return entity.info.type === 'RichTexts';
}

export function assertIsAdminRichTexts(entity: AdminEntity<string, object>): asserts entity is AdminRichTexts {
  if (entity.info.type !== 'RichTexts') {
    throw new Error('Expected info.type = RichTexts (but was ' + entity.info.type + ')');
  }
}

export type AppAdminValueItem = never;
"
`;

exports[`generateTypescriptForSchema > Strings 1`] = `
"import type { AdminClient, AdminEntity, AdminExceptionClient } from '@dossierhq/core';

export type AppAdminClient = AdminClient<AppAdminEntity, AppAdminUniqueIndexes, AppAdminExceptionClient>;

export type AppAdminExceptionClient = AdminExceptionClient<AppAdminEntity, AppAdminUniqueIndexes>;

export type AppAdminUniqueIndexes = never;

export type AppAdminEntity = AdminStrings;

export interface AdminStringsFields {
  string: string | null;
  stringList: string[] | null;
}

export type AdminStrings = AdminEntity<'Strings', AdminStringsFields, string>;

export function isAdminStrings(entity: AdminEntity<string, object>): entity is AdminStrings {
  return entity.info.type === 'Strings';
}

export function assertIsAdminStrings(entity: AdminEntity<string, object>): asserts entity is AdminStrings {
  if (entity.info.type !== 'Strings') {
    throw new Error('Expected info.type = Strings (but was ' + entity.info.type + ')');
  }
}

export type AppAdminValueItem = never;
"
`;

exports[`generateTypescriptForSchema > Unique index 1`] = `
"import type { AdminClient, AdminEntity, AdminExceptionClient } from '@dossierhq/core';

export type AppAdminClient = AdminClient<AppAdminEntity, AppAdminUniqueIndexes, AppAdminExceptionClient>;

export type AppAdminExceptionClient = AdminExceptionClient<AppAdminEntity, AppAdminUniqueIndexes>;

export type AppAdminUniqueIndexes = 'slug';

export type AppAdminEntity = AdminStrings;

export interface AdminStringsFields {
  string: string | null;
}

export type AdminStrings = AdminEntity<'Strings', AdminStringsFields, string>;

export function isAdminStrings(entity: AdminEntity<string, object>): entity is AdminStrings {
  return entity.info.type === 'Strings';
}

export function assertIsAdminStrings(entity: AdminEntity<string, object>): asserts entity is AdminStrings {
  if (entity.info.type !== 'Strings') {
    throw new Error('Expected info.type = Strings (but was ' + entity.info.type + ')');
  }
}

export type AppAdminValueItem = never;
"
`;

exports[`generateTypescriptForSchema > ValueItem (no fields) 1`] = `
"import type { AdminClient, AdminEntity, AdminExceptionClient, ValueItem } from '@dossierhq/core';

export type AppAdminClient = AdminClient<AppAdminEntity, AppAdminUniqueIndexes, AppAdminExceptionClient>;

export type AppAdminExceptionClient = AdminExceptionClient<AppAdminEntity, AppAdminUniqueIndexes>;

export type AppAdminUniqueIndexes = never;

export type AppAdminEntity = AdminValueItem;

export interface AdminValueItemFields {
  valueItem: AppAdminValueItem | null;
}

export type AdminValueItem = AdminEntity<'ValueItem', AdminValueItemFields, string>;

export function isAdminValueItem(entity: AdminEntity<string, object>): entity is AdminValueItem {
  return entity.info.type === 'ValueItem';
}

export function assertIsAdminValueItem(entity: AdminEntity<string, object>): asserts entity is AdminValueItem {
  if (entity.info.type !== 'ValueItem') {
    throw new Error('Expected info.type = ValueItem (but was ' + entity.info.type + ')');
  }
}

export type AppAdminValueItem = AdminFoo;

export type AdminFooFields = Record<never, never>;

export type AdminFoo = ValueItem<'Foo', AdminFooFields>;

export function isAdminFoo(valueItem: ValueItem<string, object> | AdminFoo): valueItem is AdminFoo {
  return valueItem.type === 'Foo';
}

export function assertIsAdminFoo(valueItem: ValueItem<string, object> | AdminFoo): asserts valueItem is AdminFoo {
  if (valueItem.type !== 'Foo') {
    throw new Error('Expected type = Foo (but was ' + valueItem.type + ')');
  }
}
"
`;

exports[`generateTypescriptForSchema > ValueTypes 1`] = `
"import type { AdminClient, AdminEntity, AdminExceptionClient, ValueItem } from '@dossierhq/core';

export type AppAdminClient = AdminClient<AppAdminEntity, AppAdminUniqueIndexes, AppAdminExceptionClient>;

export type AppAdminExceptionClient = AdminExceptionClient<AppAdminEntity, AppAdminUniqueIndexes>;

export type AppAdminUniqueIndexes = never;

export type AppAdminEntity = AdminValueTypes;

export interface AdminValueTypesFields {
  valueType: AppAdminValueItem | null;
  valueTypeFoo: AdminFoo | null;
  valueTypeFooBar: AdminBar | AdminFoo | null;
  valueTypeList: AppAdminValueItem[] | null;
  valueTypeListFoo: AdminFoo[] | null;
  valueTypeListFooBar: Array<AdminBar | AdminFoo> | null;
}

export type AdminValueTypes = AdminEntity<'ValueTypes', AdminValueTypesFields, string>;

export function isAdminValueTypes(entity: AdminEntity<string, object>): entity is AdminValueTypes {
  return entity.info.type === 'ValueTypes';
}

export function assertIsAdminValueTypes(entity: AdminEntity<string, object>): asserts entity is AdminValueTypes {
  if (entity.info.type !== 'ValueTypes') {
    throw new Error('Expected info.type = ValueTypes (but was ' + entity.info.type + ')');
  }
}

export type AppAdminValueItem = AdminBar | AdminFoo;

export interface AdminBarFields {
  string: string | null;
}

export type AdminBar = ValueItem<'Bar', AdminBarFields>;

export function isAdminBar(valueItem: ValueItem<string, object> | AdminBar): valueItem is AdminBar {
  return valueItem.type === 'Bar';
}

export function assertIsAdminBar(valueItem: ValueItem<string, object> | AdminBar): asserts valueItem is AdminBar {
  if (valueItem.type !== 'Bar') {
    throw new Error('Expected type = Bar (but was ' + valueItem.type + ')');
  }
}

export interface AdminFooFields {
  string: string | null;
}

export type AdminFoo = ValueItem<'Foo', AdminFooFields>;

export function isAdminFoo(valueItem: ValueItem<string, object> | AdminFoo): valueItem is AdminFoo {
  return valueItem.type === 'Foo';
}

export function assertIsAdminFoo(valueItem: ValueItem<string, object> | AdminFoo): asserts valueItem is AdminFoo {
  if (valueItem.type !== 'Foo') {
    throw new Error('Expected type = Foo (but was ' + valueItem.type + ')');
  }
}
"
`;

exports[`generateTypescriptForSchema > empty 1`] = `
"import type { AdminClient, AdminExceptionClient } from '@dossierhq/core';

export type AppAdminClient = AdminClient<AppAdminEntity, AppAdminUniqueIndexes, AppAdminExceptionClient>;

export type AppAdminExceptionClient = AdminExceptionClient<AppAdminEntity, AppAdminUniqueIndexes>;

export type AppAdminUniqueIndexes = never;

export type AppAdminEntity = never;

export type AppAdminValueItem = never;
"
`;

exports[`generateTypescriptForSchema published > Unique index 1`] = `
"import type { PublishedClient, PublishedEntity, PublishedExceptionClient } from '@dossierhq/core';

export type AppPublishedClient = PublishedClient<AppPublishedEntity, AppPublishedUniqueIndexes, AppPublishedExceptionClient>;

export type AppPublishedExceptionClient = PublishedExceptionClient<AppPublishedEntity, AppPublishedUniqueIndexes>;

export type AppPublishedUniqueIndexes = 'slug';

export type AppPublishedEntity = PublishedStrings;

export interface PublishedStringsFields {
  string: string | null;
}

export type PublishedStrings = PublishedEntity<'Strings', PublishedStringsFields, string>;

export function isPublishedStrings(entity: PublishedEntity<string, object>): entity is PublishedStrings {
  return entity.info.type === 'Strings';
}

export function assertIsPublishedStrings(entity: PublishedEntity<string, object>): asserts entity is PublishedStrings {
  if (entity.info.type !== 'Strings') {
    throw new Error('Expected info.type = Strings (but was ' + entity.info.type + ')');
  }
}

export type AppPublishedValueItem = never;
"
`;

exports[`generateTypescriptForSchema published > adminOnly entity type 1`] = `
"import type { PublishedClient, PublishedExceptionClient } from '@dossierhq/core';

export type AppPublishedClient = PublishedClient<AppPublishedEntity, AppPublishedUniqueIndexes, AppPublishedExceptionClient>;

export type AppPublishedExceptionClient = PublishedExceptionClient<AppPublishedEntity, AppPublishedUniqueIndexes>;

export type AppPublishedUniqueIndexes = never;

export type AppPublishedEntity = never;

export type AppPublishedValueItem = never;
"
`;

exports[`generateTypescriptForSchema published > adminOnly field 1`] = `
"import type { PublishedClient, PublishedEntity, PublishedExceptionClient } from '@dossierhq/core';

export type AppPublishedClient = PublishedClient<AppPublishedEntity, AppPublishedUniqueIndexes, AppPublishedExceptionClient>;

export type AppPublishedExceptionClient = PublishedExceptionClient<AppPublishedEntity, AppPublishedUniqueIndexes>;

export type AppPublishedUniqueIndexes = never;

export type AppPublishedEntity = PublishedFoo;

export type PublishedFooFields = Record<never, never>;

export type PublishedFoo = PublishedEntity<'Foo', PublishedFooFields, string>;

export function isPublishedFoo(entity: PublishedEntity<string, object>): entity is PublishedFoo {
  return entity.info.type === 'Foo';
}

export function assertIsPublishedFoo(entity: PublishedEntity<string, object>): asserts entity is PublishedFoo {
  if (entity.info.type !== 'Foo') {
    throw new Error('Expected info.type = Foo (but was ' + entity.info.type + ')');
  }
}

export type AppPublishedValueItem = never;
"
`;

exports[`generateTypescriptForSchema published > empty 1`] = `
"import type { PublishedClient, PublishedExceptionClient } from '@dossierhq/core';

export type AppPublishedClient = PublishedClient<AppPublishedEntity, AppPublishedUniqueIndexes, AppPublishedExceptionClient>;

export type AppPublishedExceptionClient = PublishedExceptionClient<AppPublishedEntity, AppPublishedUniqueIndexes>;

export type AppPublishedUniqueIndexes = never;

export type AppPublishedEntity = never;

export type AppPublishedValueItem = never;
"
`;

exports[`generateTypescriptForSchema published > required fields 1`] = `
"import type { PublishedClient, PublishedEntity, PublishedExceptionClient } from '@dossierhq/core';

export type AppPublishedClient = PublishedClient<AppPublishedEntity, AppPublishedUniqueIndexes, AppPublishedExceptionClient>;

export type AppPublishedExceptionClient = PublishedExceptionClient<AppPublishedEntity, AppPublishedUniqueIndexes>;

export type AppPublishedUniqueIndexes = never;

export type AppPublishedEntity = PublishedFoo;

export interface PublishedFooFields {
  string: string;
  stringList: string[];
}

export type PublishedFoo = PublishedEntity<'Foo', PublishedFooFields, string>;

export function isPublishedFoo(entity: PublishedEntity<string, object>): entity is PublishedFoo {
  return entity.info.type === 'Foo';
}

export function assertIsPublishedFoo(entity: PublishedEntity<string, object>): asserts entity is PublishedFoo {
  if (entity.info.type !== 'Foo') {
    throw new Error('Expected info.type = Foo (but was ' + entity.info.type + ')');
  }
}

export type AppPublishedValueItem = never;
"
`;
